### ISSUE TITLE
Clang-tidy: performance-move-const-arg strange failure 

### ISSUE THREAD

**Issue by JVApen** [llvm/llvm-project#61257]
Bug linked to https://discourse.llvm.org/t/some-instances-of-performance-move-const-arg-suggest-a-fix-others-do-not/69055/2?u=jvapen

In case of
````
void f(int &&);
void g()
{
    int i = 0;
    f(std::move(i));
}
````
Clang-tidy gives a warning that this move has no effect, though removing it makes the code invalid. This was identified in https://github.com/llvm/llvm-project/commit/a7f8aea71485c00db2ffb6288ae58475869048b1

As such, I believe this report is not that relevant and with std::move being elided due to recent changes, even on O0, I don't understand the performance impact it can have. 

Assuming we want to keep this case, I would expect this to be reported on the definition of the function. That would be the place where the unneeded && can be detected and both definition and declaration can be updated. 

Though, I don't believe this has nothing to do with performance, one could argue that it should not even be reported.

Possible fixes:
 - add extra diagnostic on function definition
    - optionally: remove current diagnostic
 - remove the diagnostic
 - make it configurable that next to CheckTriviallyCopyableMove, it does not report if function requires the move
 - move the trivially copyable detection to a separate check

If it is decided what option to take, this could be a good beginners bug.

**Comment by PiotrZSL** [comment]
Proper solution would be not to warn on this issue, simply if function takes rvalue, the rvalue should be.
If function would take const &, then this could be removed.
Probably it could detect implicit casts, and warn only if there is one....

```
    void f(int&&);
    `-CallExpr <line:7:5, col:19> 'void'
      |-ImplicitCastExpr <col:5> 'void (*)(int &&)' <FunctionToPointerDecay>
      | `-DeclRefExpr <col:5> 'void (int &&)' lvalue Function 0x7ff85a52a140 'f' 'void (int &&)'
      `-CallExpr <col:7, col:18> 'typename std::remove_reference<int &>::type':'int' xvalue
        |-ImplicitCastExpr <col:7, col:12> 'typename std::remove_reference<int &>::type &&(*)(int &) noexcept' <BuiltinFnToFnPtr>
        | `-DeclRefExpr <col:7, col:12> '<builtin fn type>' Function 0x7ff85a52ab50 'move' 'typename std::remove_reference<int &>::type &&(int &) noexcept' (FunctionTemplate 0x5563099b8408 'move')
        `-DeclRefExpr <col:17> 'int' lvalue Var 0x7ff85a52a300 'i' 'int'

    void f(const int&);
    `-CallExpr <line:7:5, col:19> 'void'
      |-ImplicitCastExpr <col:5> 'void (*)(const int &)' <FunctionToPointerDecay>
      | `-DeclRefExpr <col:5> 'void (const int &)' lvalue Function 0x7efd26bf1110 'f' 'void (const int &)'
      `-ImplicitCastExpr <col:7, col:18> 'const typename std::remove_reference<int &>::type':'const int' xvalue <NoOp>
        `-CallExpr <col:7, col:18> 'typename std::remove_reference<int &>::type':'int' xvalue
          |-ImplicitCastExpr <col:7, col:12> 'typename std::remove_reference<int &>::type &&(*)(int &) noexcept' <BuiltinFnToFnPtr>
          | `-DeclRefExpr <col:7, col:12> '<builtin fn type>' Function 0x7efd26bf1b50 'move' 'typename std::remove_reference<int &>::type &&(int &) noexcept' (FunctionTemplate 0x56496135f1c8 'move')
          `-DeclRefExpr <col:17> 'int' lvalue Var 0x7efd26bf12d0 'i' 'int'

    void f(int);
    `-CallExpr <line:7:5, col:19> 'void'
      |-ImplicitCastExpr <col:5> 'void (*)(int)' <FunctionToPointerDecay>
      | `-DeclRefExpr <col:5> 'void (int)' lvalue Function 0x564e30204c40 'f' 'void (int)'
      `-ImplicitCastExpr <col:7, col:18> 'typename std::remove_reference<int &>::type':'int' <LValueToRValue>
        `-CallExpr <col:7, col:18> 'typename std::remove_reference<int &>::type':'int' xvalue
          |-ImplicitCastExpr <col:7, col:12> 'typename std::remove_reference<int &>::type &&(*)(int &) noexcept' <BuiltinFnToFnPtr>
          | `-DeclRefExpr <col:7, col:12> '<builtin fn type>' Function 0x564e30205680 'move' 'typename std::remove_reference<int &>::type &&(int &) noexcept' (FunctionTemplate 0x564e2ed89ff8 'move')
          `-DeclRefExpr <col:17> 'int' lvalue Var 0x564e30204e00 'i' 'int'
```
